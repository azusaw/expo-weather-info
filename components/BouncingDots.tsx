import React, { useEffect, useRef } from "react";
import { Animated, View, StyleSheet } from "react-native";
import Colors from "@/constants/Colors";

const BouncingDots = () => {
  const dots = [
    useRef(new Animated.Value(0)).current,
    useRef(new Animated.Value(0)).current,
    useRef(new Animated.Value(0)).current,
  ];

  const colors = [
    Colors.secondary.light,
    Colors.secondary.default,
    Colors.secondary.dark,
  ];

  useEffect(() => {
    dots.forEach((dot, index) => {
      Animated.loop(
        Animated.sequence([
          Animated.delay(index * 150),
          Animated.timing(dot, {
            toValue: -10,
            duration: 300,
            useNativeDriver: true,
          }),
          Animated.timing(dot, {
            toValue: 0,
            duration: 300,
            useNativeDriver: true,
          }),
        ]),
      ).start();
    });
  }, []);

  return (
    <View style={styles.container}>
      {dots.map((dot, idx) => (
        <Animated.View
          key={idx}
          style={[
            styles.dot,
            { backgroundColor: colors[idx], transform: [{ translateY: dot }] },
          ]}
        />
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  dot: {
    width: 20,
    height: 20,
    marginHorizontal: 5,
    borderRadius: 5,
  },
});

export default BouncingDots;
