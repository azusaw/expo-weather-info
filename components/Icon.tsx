import React from "react";
import Svg from "react-native-svg";
import { IconPaths } from "@/constants/IconPaths";

export enum IconType {
  Drop = "drop",
  Rain = "rain",
  Wind = "wind",
}

type IconProps = {
  name: IconType;
  size?: number;
  color?: string;
};

const getIconPath = (name: IconType) => {
  switch (name) {
    case IconType.Drop:
      return IconPaths.drop;
    case IconType.Rain:
      return IconPaths.rain;
    case IconType.Wind:
      return IconPaths.wind;
    default:
      return <></>;
  }
};

const SvgIcon = ({ name, size = 100, color = "#888" }: IconProps) => (
  <Svg viewBox="0 0 570 570" width={size} height={size} fill={color}>
    {getIconPath(name)}
  </Svg>
);

export default SvgIcon;
