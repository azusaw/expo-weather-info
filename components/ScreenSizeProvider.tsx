import { createContext, useContext, PropsWithChildren } from "react";
import { useScreenSize } from "@/hooks/useScreenSize";

export const ScreenSizeContext = createContext<{
  isSmall: boolean;
  height: number;
  width: number;
}>({
  isSmall: false,
  height: 0,
  width: 0,
});

export const ScreenSizeProvider = ({ children }: PropsWithChildren) => {
  const size = useScreenSize();
  return (
    <ScreenSizeContext.Provider value={size}>
      {children}
    </ScreenSizeContext.Provider>
  );
};

export const useScreenSizeContext = () => useContext(ScreenSizeContext);
