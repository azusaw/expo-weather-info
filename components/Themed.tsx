import { Text as DefaultText, View as DefaultView } from "react-native";
import Colors from "@/constants/Colors";

type ThemeProps = {
  lightColor?: string;
  darkColor?: string;
};

type FontProps = {
  size?: number;
  weight?: 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900;
  color?: string;
};

export type TextProps = ThemeProps & FontProps & DefaultText["props"];
export type ViewProps = ThemeProps & DefaultView["props"];

const getFontFamily = (fontWeight: number) => {
  switch (fontWeight) {
    case 100:
      return "Montserrat_100Thin";
    case 200:
      return "Montserrat_200ExtraLight";
    case 300:
      return "Montserrat_300Light";
    case 400:
      return "Montserrat_400Regular";
    case 500:
      return "Montserrat_500Medium";
    case 600:
      return "Montserrat_600SemiBold";
    case 700:
      return "Montserrat_700Bold";
    case 800:
      return "Montserrat_800ExtraBold";
    case 900:
      return "Montserrat_900Black";
    default:
      return "Montserrat_400Regular";
  }
};

export const Text = (props: TextProps) => {
  const {
    style,
    size = 24,
    weight = 400,
    color = Colors.text.dark,
    ...otherProps
  } = props;
  const fontFamily = getFontFamily(weight);

  return (
    <DefaultText
      style={[{ fontFamily, fontSize: size, color }, style]}
      {...otherProps}
    />
  );
};

export const View = (props: ViewProps) => {
  const { style, lightColor, darkColor, ...otherProps } = props;
  const backgroundColor = "inherit";

  return <DefaultView style={[{ backgroundColor }, style]} {...otherProps} />;
};
