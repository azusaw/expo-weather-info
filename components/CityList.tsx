import React, { useState } from "react";
import { StyleSheet, TouchableOpacity } from "react-native";
import CityLocations from "@/assets/json/city-location.json";
import Colors from "@/constants/Colors";
import { Text, View } from "./Themed";
import { useLocationStore } from "@/store/useLocationStore";

const CityList = React.memo(({ onClose }: { onClose: () => void }) => {
  const { setLocation } = useLocationStore();
  const [selectedCityIndex, setSelectedCityIndex] = useState<number>();

  const onSelect = async (index: number) => {
    setSelectedCityIndex(index); // for showing active color
    await setLocation(CityLocations[index]);
    onClose();
  };

  return (
    <View style={styles.container}>
      <View style={styles.menuBox}>
        {CityLocations.map(({ name }, index) => (
          <TouchableOpacity key={name} onPress={() => onSelect(index)}>
            <View
              style={
                index === selectedCityIndex
                  ? styles.activeItem
                  : styles.menuItem
              }
            >
              <Text size={18}>{name}</Text>
            </View>
            {index < CityLocations.length - 1 && <View style={styles.border} />}
          </TouchableOpacity>
        ))}
      </View>
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    position: "relative",
    marginHorizontal: "auto",
    width: "100%",
    maxWidth: 400,
  },
  menuBox: {
    position: "absolute",
    bottom: 50,
    right: 5,
    width: 220,
    borderRadius: 5,
    backgroundColor: "#fff",
    overflow: "hidden",
  },
  menuItem: {
    padding: 15,
    borderRadius: 5,
  },
  activeItem: {
    padding: 15,
    backgroundColor: Colors.primary.default,
    borderLeftWidth: 6,
    borderLeftColor: Colors.secondary.default,
  },
  border: {
    height: 1,
    width: "100%",
    backgroundColor: Colors.primary.default,
  },
});

export default CityList;
