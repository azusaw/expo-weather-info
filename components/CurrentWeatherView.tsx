import React from "react";
import dayjs from "dayjs";
import "dayjs/locale/en";
import { StyleSheet } from "react-native";
import { Text, View } from "@/components/Themed";
import { getWeatherDetailFromCode } from "@/libs/getWeatherDetailFromCode";
import { CurrentWeather } from "@/types";
import Icon, { IconType } from "@/components/SvgIcon";
import Colors from "@/constants/Colors";
import { useScreenSizeContext } from "@/components/ScreenSizeProvider";

const CurrentWeatherView = ({
  data,
  siteName = "Your Location",
}: {
  data: CurrentWeather;
  siteName?: string;
}) => {
  const { isSmall } = useScreenSizeContext();
  const infoItems: { label: string; icon: IconType; data: string }[] = [
    { label: "Rain", icon: IconType.Rain, data: `${data.rain}mm` },
    { label: "Wind", icon: IconType.Wind, data: `${data.windSpeed}km/h` },
    { label: "Humidity", icon: IconType.Drop, data: `${data.humidity}%` },
  ];
  return (
    <View style={styles.container}>
      <Text size={isSmall ? 32 : 36} weight={700} style={styles.siteName}>
        {siteName}
      </Text>
      <Text size={isSmall ? 14 : 16} weight={300} style={styles.date}>
        {dayjs(data.time).format("DD MMM ddd, hh A")}
      </Text>
      <Text size={isSmall ? 20 : 24} weight={300}>
        {getWeatherDetailFromCode(data.weatherCode).description}
      </Text>
      <Text size={isSmall ? 80 : 100} weight={400} style={styles.temperature}>
        {data.temperature}
        <View style={[styles.unit, isSmall && styles.unitSmall]}>
          <View
            style={[styles.unitCircle, isSmall && styles.unitCircleSmall]}
          />
        </View>
      </Text>
      <View style={[styles.card, isSmall && styles.cardSmall]}>
        {infoItems.map(({ label, icon, data }) => (
          <View key={label} style={styles.cardItem}>
            <Icon
              name={icon}
              size={isSmall ? 30 : 40}
              color={Colors.primary.default}
            />
            <Text size={isSmall ? 14 : 18} style={{ marginTop: 20 }}>
              {data}
            </Text>
            <Text size={isSmall ? 10 : 14} style={{ marginTop: 10 }}>
              {label}
            </Text>
          </View>
        ))}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    backgroundColor: "inherit",
  },
  siteName: {
    marginBottom: 20,
  },
  date: {
    backgroundColor: Colors.primary.dark,
    paddingHorizontal: 20,
    paddingVertical: 5,
    borderRadius: 20,
    color: "#fff",
    marginBottom: 24,
  },
  temperature: {
    position: "relative",
  },
  unit: {
    position: "absolute",
    top: 30,
    right: -30,
  },
  unitSmall: {
    position: "absolute",
    top: 20,
    right: -20,
  },
  unitCircle: {
    borderRadius: 10,
    backgroundColor: "transparent",
    borderColor: Colors.text.light,
    justifyContent: "center",
    alignItems: "center",
    width: 20,
    height: 20,
    borderWidth: 4,
  },
  unitCircleSmall: {
    width: 15,
    height: 15,
    borderWidth: 4,
  },
  card: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-evenly",
    alignItems: "center",
    paddingVertical: 10,
    paddingHorizontal: 20,
    marginTop: 8,
    borderRadius: 10,
    backgroundColor: Colors.primary.dark,
  },
  cardSmall: {
    paddingVertical: 5,
    paddingHorizontal: 15,
  },
  cardItem: {
    alignItems: "center",
    padding: 15,
  },
});

export default CurrentWeatherView;
