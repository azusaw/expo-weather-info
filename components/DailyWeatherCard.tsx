import React from "react";
import { StyleSheet } from "react-native";
import { DailyWeather } from "@/types";
import { Text, View } from "@/components/Themed";
import WeatherImage from "@/components/WeatherImage";
import dayjs from "dayjs";

const DailyWeatherCard = ({
  time,
  weatherCode,
  temperatureMin,
  temperatureMax,
}: DailyWeather) => (
  <View style={styles.card}>
    <Text style={styles.date}>{dayjs(time).format("DD ddd")}</Text>
    <WeatherImage weatherCode={weatherCode} />
    <Text style={styles.temperature}>
      {temperatureMin}
      <Text style={styles.slash}>/</Text>
      {temperatureMax}
      <Text style={styles.suffix}>â„ƒ</Text>
    </Text>
  </View>
);
export default DailyWeatherCard;

const styles = StyleSheet.create({
  card: {
    minWidth: 200,
    alignItems: "center",
    backgroundColor: "#fff",
    elevation: 3,
    marginLeft: 10,
    marginRight: 10,
    padding: 20,
    borderRadius: 10,
  },
  date: {
    textAlign: "center",
    fontSize: 24,
  },
  temperature: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    marginTop: 12,
    fontSize: 20,
    fontWeight: 100,
  },
  slash: {
    marginLeft: 10,
    marginRight: 10,
    fontSize: 24,
  },
  suffix: {
    marginLeft: 10,
    fontSize: 24,
  },
});
