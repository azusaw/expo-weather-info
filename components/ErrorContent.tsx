import { Button, StyleSheet } from "react-native";
import { Text, View } from "@/components/Themed";
import Colors from "@/constants/Colors";

const ErrorContent = ({
  message,
  onRetry,
}: {
  message: string;
  onRetry: () => void;
}) => (
  <View>
    <Text size={42} weight={700} color={Colors.text.light}>
      Sorry...
    </Text>
    <Text size={22} color={Colors.text.light} style={styles.description}>
      {message}
    </Text>
    <Button title="Retry" onPress={onRetry} color={Colors.primary.dark} />
  </View>
);

const styles = StyleSheet.create({
  description: {
    marginTop: 20,
    marginBottom: 40,
  },
});

export default ErrorContent;
