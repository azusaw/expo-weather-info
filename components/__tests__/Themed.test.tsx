import { render } from "@testing-library/react-native";
import { Text, View } from "@/components/Themed";

describe("Text", () => {
  test("it renders Text successfully", () => {
    const { toJSON } = render(<Text>Hello</Text>);
    expect(toJSON()).toMatchSnapshot();
  });

  [
    [100, "Montserrat_100Thin"],
    [200, "Montserrat_200ExtraLight"],
    [300, "Montserrat_300Light"],
    [400, "Montserrat_400Regular"],
    [500, "Montserrat_500Medium"],
    [500, "Montserrat_500Medium"],
    [600, "Montserrat_600SemiBold"],
    [700, "Montserrat_700Bold"],
    [800, "Montserrat_800ExtraBold"],
    [900, "Montserrat_900Black"],
    [undefined, "Montserrat_400Regular"],
  ].forEach(([weight, font]) =>
    test(`it renders font weight ${weight}`, () => {
      // @ts-ignore
      const { getByText } = render(<Text weight={weight}>Hello</Text>);
      // @ts-ignore
      expect(getByText("Hello")).toHaveStyle({ fontFamily: font });
    }),
  );
});

describe("View", () => {
  test("it renders View successfully", () => {
    const { toJSON } = render(<View>Hello</View>);
    expect(toJSON()).toMatchSnapshot();
  });
});
