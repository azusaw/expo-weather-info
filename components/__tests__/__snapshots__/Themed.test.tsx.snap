// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text it renders Text successfully 1`] = `
<Text
  style={
    [
      {
        "color": "#888",
        "fontFamily": "Montserrat_400Regular",
        "fontSize": 24,
      },
      undefined,
    ]
  }
>
  Hello
</Text>
`;

exports[`View it renders Text successfully 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "inherit",
      },
      undefined,
    ]
  }
>
  Hello
</View>
`;

exports[`View it renders View successfully 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "inherit",
      },
      undefined,
    ]
  }
>
  Hello
</View>
`;
